internal/vault/vault_test.go

```diff
@@ -1169,3 +1170,57 @@ func TestRequestTokenWithAppRoleRef(t *testing.T) {
 		})
 	}
 }
+
+type testNamespacedRequestT struct {
+	expectedNamespace string
+
+	issuer *cmapi.VaultIssuer
+}
+
+func TestNamespacedRequest(t *testing.T) {
+	tests := map[string]testNamespacedRequestT{
+		"request should include the namespace when present on the issuer": {
+			expectedNamespace: "test-namespace",
+			issuer:            &cmapi.VaultIssuer{Namespace: "test-namespace"},
+		},
+		"request should not include the namespace header when namespace field is empty": {
+			expectedNamespace: "",
+			issuer:            &cmapi.VaultIssuer{},
+		},
+		"request should not include the namespace header when passed a nil issuer": {
+			expectedNamespace: "",
+			issuer:            nil,
+		},
+	}
+
+	var ns string
+	handler := func(w http.ResponseWriter, req *http.Request) {
+		ns = req.Header.Get("X-Vault-Namespace")
+	}
+	server := httptest.NewServer(http.HandlerFunc(handler))
+	defer server.Close()
+
+	config := vault.DefaultConfig()
+	config.Address = server.URL
+
+	vaultClient, err := vault.NewClient(config)
+	if err != nil {
+		t.Errorf("error initializing vault client: %s", err)
+		t.FailNow()
+	}
+
+	client := vaultClientWrapper{Client: vaultClient}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			_, err = client.NamespacedRawRequest(test.issuer, client.NewRequest("GET", "/some/test/path"))
+			if err != nil {
+				t.Errorf("error making request: %s", err)
+				t.FailNow()
+			}
+			if ns != test.expectedNamespace {
+				t.Errorf("namespace header does not match expected value, exp=%s got=%s", test.expectedNamespace, ns)
+			}
+		})
+	}
+}
```

Comment:
I'm not convinced we should be  testing for the presence of the `X-Vault-Namespace` header. 
That's an implementation detail of the Vault server. 
What if Hashicorp introduce a slightly different Header or some other mechanism for configuring the namespace?

The vault SDK client has a `Namespace()` method which tells us which namespace it's going to use, and in 
https://github.com/cert-manager/cert-manager/pull/5591 I explored a slightly different way of testing this and implemented the namespace and non-namespace client following a pattern I found in Hashicorp Nomad. 

Please tell me what you think of that approach and if you both prefer this PR, I'll concede and approve this PR.

Reply:
I prefer your approach in #5591. 

I think based on the points you raised, #5591 is clearer and more consistent. It also eliminates ambiguity around when to use or not use a namespaced request. In your implementation you just use the default client for everything except sys calls.

I also agree with your point about this test being tightly coupled to Vault's implementation. And it's better to just check that the client has the expected namespace set. I did like being able to check what was actually received on the server side, but that's a better test for the vault client itself.

I'm all for closing this PR in favor of #5591 🚀